### ISSUE TITLE
[mojo-stdlib] Use `Arc` in manual RefCounted structs in `object.mojo`

### ISSUE THREAD

**Issue by JoeLoser** [modularml/mojo#2030]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

There are several "manual refcounted" structs in `object.mojo` such as [this one](https://github.com/modularml/mojo/blob/3eb006a3b21e7505b9f48fb00dd5f1cbcbe54848/stdlib/src/builtin/object.mojo#L79).  Now that we have an atomic ref-counted type (spelled `Arc`) [here](https://github.com/modularml/mojo/blob/3eb006a3b21e7505b9f48fb00dd5f1cbcbe54848/stdlib/src/memory/_arc.mojo), we should use it!

### What is your motivation for this change?

This will simplify the code and push on the API we have for `Arc` right now which is a "private type" (undocumented since the filename is prefixed with an underscore).  Once we push more use of it in the standard library, I'd be up for renaming the file from `_arc.mojo` to `arc.mojo`.

**Comment by jayzhan211** [comment]
How do we call the method of T for `Arc[T]`? For example, List has the method `append`, but it doesn't seem there is any trivial way to do `append` for `Arc[T]`.

**Comment by LJ-9801** [comment]
So should we refactor both the `struct _RefCountedListRef `  and `struct _RefCountedAttrsDictRef` so that it uses the `Arc`?

**Comment by jayzhan211** [comment]
I think they should be deprecated and replaced with Arc[T]

**Comment by JoeLoser** [comment]
I'd recommend just replacing them in-place to internally use an `Arc` as a data member.  There's no need to go through a deprecation of private types (ones prefixed with an underscore such as `_RefCountedListRef`).

**Comment by JoeLoser** [comment]
> How do we call the method of T for `Arc[T]`? For example, List has the method `append`, but it doesn't seem there is any trivial way to do `append` for `Arc[T]`.

`Arc` has a [refitem function](https://github.com/modularml/mojo/blob/daace814835ae7e20a65376a18244c479e0a2647/stdlib/src/memory/_arc.mojo#L140) to get the underlying type it holds.  In this case, it returns a `Reference` to the `T`, so you'll need to do `[]` to dereference the thing to get a `T`.  Does that make sense?

**Comment by jayzhan211** [comment]
As you mentioned that `[]` is equivalent to calling `__refitem__`. Is there any way that we can know the corresponding function for the built-in operator?  I had read the function `__refitem__`, but I didn't notice it can be called by `[]`. And then, I found accessing with `.__refitem__` non-trivial.

**Comment by sepeth** [comment]
This seems done?

**Comment by JoeLoser** [comment]
Yep, this is done via [https://github.com/modularml/mojo/pull/2231](https://github.com/modularml/mojo/pull/2231) and [https://github.com/modularml/mojo/pull/2146](https://github.com/modularml/mojo/pull/2146) it looks like.  I'll close this issue.  Thanks for looking into it!
